test/metabase/models/task_history_test.clj

```diff
@@ -71,3 +71,33 @@
         (task-history/cleanup-task-history! 100)
         (is (= #{task-1 task-2}
                (set (map :task (t2/select TaskHistory)))))))))
+
+(deftest with-task-history-test
+  (mt/with-model-cleanup [:model/TaskHistory]
+    (testing "success path:"
+      (let [task-name (mt/random-name)]
+        (testing "task history is created before executing the body"
+          (task-history/with-task-history {:task task-name}
+            (is (=? {:status     :started
+                     :started_at (mt/malli=? some?)
+                     :ended_at   (mt/malli=? nil?)
+                     :duration   (mt/malli=? nil?)}
+                    (t2/select-one :model/TaskHistory :task task-name)))))
+        (testing "when the task is done, updates status and duration correctly"
+          (is (=? {:status     :success
+                   :started_at (mt/malli=? some?)
+                   :ended_at   (mt/malli=? some?)
+                   :duration   (mt/malli=? int?)}
```

Comment:
```suggestion
                   :duration   (mt/malli=? nat-int?)}
```

You can only strenghten this if you move to `System/nanoTime` though, system clock *can* go backwards.